### PR TITLE
feat(core): allow options for action date & date range

### PR DESCRIPTION
**Pull request by P0ppoff**
## Checklist

Definition Of Review

-   [x] Own code review done
-   [x] Write message in teams channel

Definition Of Mergeable

-   [ ] 2 approves
-   [ ] 1 functional review - US has been tested
-   [ ] Every comment is handled - blocking ones have been resolved by reviewer
-   [ ] Design OK
-   [ ] Can be tested by POs
-   [ ] PR was introduced during daily meeting

### CODE REVIEW

File: apps/core/src/domain/actions/formatDateRangeAction.ts [hunk 1]
```diff
@@ -14,47 +22,68 @@ export default function (): IActionsListFunction<{auto: true; format: false}> {
...
+                            to: moment.unix(numberValTo).format(params['Imposing format'])
+                        }
+                    };
+                }
+
+                let options: Intl.DateTimeFormatOptions = {};
+                try {
+                    options = JSON.parse(params.Dynamic ?? {});
+                } catch (e) {
+                    // TODO: rise error to inform user without break app
```
Comment (P0ppoff) [inline apps/core/src/domain/actions/formatDateRangeAction.ts:71]: **note:** on le fera dans ce ticket https://aristid.atlassian.net/browse/XSTREAM-606

File: apps/core/src/domain/actions/formatDateAction.spec.ts [hunk 2]
```diff
@@ -9,31 +9,121 @@ describe('formatDateAction', () => {
...
+                        {
+                            ...ctx,
+                            lang: 'ar-EG'
+                        }
+                    )
+                ).values[0].value
+            ).toBe('٢٨ فبراير ٣٧ في ١١ م');
+        });
+
+        test.skip.each(['{', '{withoutDoubleQuote: true}', '', '{"params1": "long", "param2": "too many coma",}'])(
```
Comment (P0ppoff) [inline apps/core/src/domain/actions/formatDateAction.spec.ts:61]: **note:** J'ai mis un TODO dans le code pour dire qu'il fallait gérer les erreurs, il y a un JIRA dessus.

File: apps/core/src/domain/actions/formatDateAction.ts [hunk 3]
```diff
@@ -2,9 +2,18 @@
 // This file is released under LGPL V3
 // License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
 import moment from 'moment';
-import {ActionsListIOTypes, IActionsListFunction} from '../../_types/actionsList';
+import {ActionsListIOTypes, IActionsListFunction, IActionsListFunctionResult} from '../../_types/actionsList';
+import {Errors} from '../../_types/errors';
 
-export default function (): IActionsListFunction<{auto: true; format: false}> {
+// TODO: find right place to store this
```
Comment (philippechevieux) [inline apps/core/src/domain/actions/formatDateAction.ts:8]: **todo:** Ici c'est bien 😉 

```suggestion

```

**Review by philippechevieux — APPROVED** [review]

File: apps/admin/src/components/attributes/EditAttribute/EditAttributeTabs/ActionsListTab/ALCCard/ActionContent/Param/Param.tsx [hunk 4]
```diff
@@ -74,6 +71,7 @@ function Param({param, actionId, changeParam, setBlockCard, index}: IParamProps)
                         onChange={_onChange}
                         onFocus={_onFocus}
                         onBlur={_onBlur}
+                        rows={7}
```
Comment (renaudAmsellem) [inline apps/admin/src/components/attributes/EditAttribute/EditAttributeTabs/ActionsListTab/ALCCard/ActionContent/Param/Param.tsx:74]: **polish:** magic number?

File: apps/core/src/domain/actions/formatDateAction.spec.ts [hunk 5]
```diff
@@ -9,31 +9,121 @@ describe('formatDateAction', () => {
...
+                expect(result.errors[0]).toEqual({
+                    attributeValue: {value: Dynamic},
+                    errorType: 'FORMAT_ERROR',
+                    message: 'Options of FormatDateAction are invalid JSON. Use `{}` empty option instead.'
+                });
+            }
+        );
+    });
+
+    test('Imposing format', async () => {
```
Comment (renaudAmsellem) [inline apps/core/src/domain/actions/formatDateAction.spec.ts:82]: **polish:** 
```suggestion
    test('Enforce format', async () => {
```
Je pense que le traduction est meilleure. Pas 100% sur.

File: apps/core/src/domain/actions/formatDateAction.spec.ts [hunk 6]
```diff
@@ -9,31 +9,121 @@ describe('formatDateAction', () => {
...
-        expect((await action([{value: 2119477320}], {auto: 'true'}, {...ctx, lang: 'fr-FR'})).values[0].value).toBe(
-            '28/02/2037 23:42:00'
-        );
-        expect((await action([{value: 2119477320}], {auto: 'true'}, {...ctx, lang: 'ar-EG'})).values[0].value).toBe(
-            '٢٨‏/٢‏/٢٠٣٧، ١١:٤٢:٠٠ م'
-        );
+    describe('edge cases', () => {
+        test('should display default localString on neither params provided', async () => {
+            expect((await action([{value: testingDate}], {}, ctx)).values[0].value).toBe(
+                (await action([{value: testingDate}], {Dynamic: '{}'}, ctx)).values[0].value
```
Comment (renaudAmsellem) [inline apps/core/src/domain/actions/formatDateAction.spec.ts:99]: **question:** Je ne comprends pas ce que fait ce test.

File: apps/core/src/domain/actions/formatDateRangeAction.ts [hunk 7]
```diff
@@ -3,9 +3,17 @@
...
 
-export default function (): IActionsListFunction<{auto: true; format: false}> {
+const defaultValueDynamicParam = `{
+  "weekday": "long",
+  "month": "long",
+  "day": "numeric",
+  "year": "numeric"
+}`;
+
+export default function (): IActionsListFunction<{Dynamic: false; 'Imposing format': false}> {
```
Comment (renaudAmsellem) [inline apps/core/src/domain/actions/formatDateRangeAction.ts:16]: **thought:** On ne peut pas lui donner un nom de variable ? Plutôt qu'une chaine de caractère avec des majuscules, minuscules et des espaces :o

File: apps/core/src/domain/actions/formatDateRangeAction.spec.ts [hunk 8]
```diff
@@ -10,64 +10,126 @@ describe('formatDateRangeAction', () => {
...
+                    await action(
+                        [{value: testingRangeDate}],
+                        {Dynamic},
+                        {
+                            ...ctx,
+                            lang: 'ar-EG'
+                        }
+                    )
+                ).values[0].value
+            ).toEqual({from: '٢٨ فبراير ٣٧ في ١١:٤٢ م', to: '٢٨ فبراير ٣٧ في ١١:٤٣ م'});
```
Comment (renaudAmsellem) [inline apps/core/src/domain/actions/formatDateRangeAction.spec.ts:60]: **nitpick:** je ne suis pas fan des tests dans des langues autant étrangères. Si jamais ce test vient à échouer, comment savoir ce qui ne va pas.
Doubler un test avec une langue de cette manière : OK ; écrire plusieurs tests dont le résultat est incompréhensible = KO à mes yeux.

File: apps/core/src/domain/actions/formatDateRangeAction.spec.ts [hunk 9]
```diff
@@ -10,64 +10,126 @@ describe('formatDateRangeAction', () => {
...
+        });
+
+        test.skip.each(['{', '{withoutDoubleQuote: true}', '', '{"params1": "long", "param2": "too many coma",}'])(
```
Comment (renaudAmsellem) [inline apps/core/src/domain/actions/formatDateRangeAction.spec.ts:63]: **nitpick:** skip?

File: apps/admin/src/components/attributes/EditAttribute/EditAttributeTabs/ActionsListTab/ALCCard/ActionContent/Param/Param.tsx (hunk 4 above)
Comment (P0ppoff) [reply apps/admin/src/components/attributes/EditAttribute/EditAttributeTabs/ActionsListTab/ALCCard/ActionContent/Param/Param.tsx:74]: oui magic number totalement, je pourrais éventuellement mettre 
```suggestion
                        rows={enougthRowToSeeDefaultValues}
```

File: apps/core/src/domain/actions/formatDateAction.spec.ts (hunk 6 above)
Comment (P0ppoff) [reply apps/core/src/domain/actions/formatDateAction.spec.ts:99]: le test s'appelle `should display default localString on neither params provided`, donc c'est la même chose que de ne pas mettre le paramètre que de le mettre à un objet `{}`, je ne savais pas trop comment l'écrire

File: apps/core/src/domain/actions/formatDateRangeAction.spec.ts (hunk 9 above)
Comment (P0ppoff) [reply apps/core/src/domain/actions/formatDateRangeAction.spec.ts:63]: oui ça rejoins ce commentaire : https://github.com/leav-solutions/leav-engine/pull/536#discussion_r1713848488
et celui là aussi : https://github.com/leav-solutions/leav-engine/pull/536#discussion_r1713905777

File: apps/core/src/domain/actions/formatDateRangeAction.ts (hunk 7 above)
Comment (P0ppoff) [reply apps/core/src/domain/actions/formatDateRangeAction.ts:16]: Capture d’écran 2024-08-14 à 14 53 01

File: apps/core/src/domain/actions/formatDateRangeAction.spec.ts (hunk 8 above)
Comment (P0ppoff) [reply apps/core/src/domain/actions/formatDateRangeAction.spec.ts:60]: oki j'avais fait ça dans une autre PR mais je met un truc en anglais un en français, ce que j'aime bien avec le koréen et l'arabe c'est que l'on voit direct que c'est très dépend de la langue, ça illustre bien le changement de langue mais je te rejoins sur le point très pertinent que si ça plante bah je ne parle pas koréen ni arabe donc pas simple de corriger. Je change ça de suite.

Comment (P0ppoff) [reply apps/core/src/domain/actions/formatDateRangeAction.spec.ts]: j'ai poussé, je te laisse voir si ça suffit et fermer le thread

File: apps/admin/src/components/attributes/EditAttribute/EditAttributeTabs/ActionsListTab/ALCCard/ActionContent/Param/Param.tsx (hunk 4 above)
Comment (P0ppoff) [reply apps/admin/src/components/attributes/EditAttribute/EditAttributeTabs/ActionsListTab/ALCCard/ActionContent/Param/Param.tsx:74]: on l'enlève

File: apps/core/src/domain/actions/formatDateAction.spec.ts (hunk 5 above)
Comment (P0ppoff) [reply apps/core/src/domain/actions/formatDateAction.spec.ts:70]: Oki on part sur `{localized: false; universal: false}`
